text/0655-enhanced-l1.md

```diff
@@ -0,0 +1,237 @@
+
+## Resource interface
+
+One of the main pain points of working with L1s is that they can’t be used when L2s are expected. This creates a disconnect between L1s and L2s where using both in one CDK app is difficult. To resolve this we need to find a way to pass L1s in where L2s are expected and vice versa, enhancing flexibility and allow developers to more easily adopt L1s.
+
+### Current behavior
+
+Most L2s currently only accept object references. This is done by hand writing resource interfaces in all L2 constructs and that interface is used whenever a construct is referenced anywhere. This gives developers a richer interaction with the referenced object as they don’t have to worry about which attribute they need to reference and they can also utilize logic encapsulated by the target object.
+
+e.x in the Bucket L2 construct
+```
+export interface IBucket extends IResource {
+    readonly bucketName: ...
+    readonly bucketArn: ...
+    ...
+    addToResourcePolicy(permission: iam.PolicyStatement) ...
+    grantRead(identity: iam.IGrantable, objectsKeyPattern?: any) ...
+    ...
+}
+```
+Anytime the bucket needs to be referenced in another construct it will take in IBucket
+```
+export interface CodePipelineProps {
+    ...
+    readonly artifactBucket?: s3.IBucket;
+}
+```
+
+L1s on the other hand don’t have any object references and just take strings to refer to other objects (names, ARNs, ids etc). They also don’t implement resource interfaces, thus cannot be passed where L2s are expected.
+
+### Proposed solution
+
+To make L1s and L2s interoperable we first need to emit a minimal interface in the L1s file with a change to the generation script. This interface should include the refs attributes for that resource (marked as `primaryIdentifiers` in the schema) and any arn attributes. 
+
+e.x for the S3Bucket L1
+```
+export interface ICfnBucket {
+    attrBucketName ...
+    attrArn ...
+}
+```
+In the L2’s resource interface we can extend this L1 interface. We can then modify all references of the L2 interface to now take in the new L1 interface. Any location that accepted `IBucket` would now accept `ICfnBucket`
+
+e.x in the Bucket L2
+```
+export interface IBucket extends ICfnBucket, IResource {
+...
+}
+```
+Any location that accepted `IBucket` would now accept `ICfnBucket`
+```
+export interface CodePipelineProps {
+    ...
+    readonly artifactBucket?: s3.ICfnBucket;
+}
+```
+If those locations use a functionality from the L2 interface, we can do an upconversion to an `IBucket` by implementing an API that converts an `ICfn<Resource>` into an `I<Resource>`. We do a similar operation when a resource that isn’t created in the current CDK stack is referenced. This needs to done to maintain backwards compatibility and not break existing customers.
```

Comment:
I am wondering if perhaps it would be sufficient to just add `fromCfnXxx` to all L2s instead of juggling all of these interfaces.

These conversion static methods would be straightforward to implement:

```js
public static fromCfnBucket(bucket: CfnBucket): IBucket {
  return this.fromBucketArn(bucket.attrBucketArn);
}
```

Indeed, it won't allow users to directly pass an L1 everywhere an `IBucket` is needed but it feels like this would be good enough to improve the ergonomics without too many breaking changes.

Reply:
Bucket L2 already has a `fromCfnBucket` that implements what you mentioned but I'm not sure how many other L2s have something similar or what the effort to add the missing ones would be.

Reply:
It should be easy to enforce the `fromCfnXxx` static methods using [awslint](https://github.com/aws/aws-cdk/tree/main/packages/awslint) so it's semi-manual.